### DISCUSSION TITLE
Parabolic Variance/Deviation, initial version available 2023-11

### DISCUSSION THREAD

**Discussion by aewallin** [aewallin/allantools#148] [Announcements]
I've pushed updates for a first version of pdev() Parabolic Deviation, following the papers Vernotte2015 and Vernotte2020.
https://github.com/aewallin/allantools/blob/master/allantools/allantools.py#L427

The code runs quite slowly, because of nested loops/sums. If anyone has good ideas on how to make it faster these are very welcome!

![2023-11-25_pdev_nbs14](https://github.com/aewallin/allantools/assets/909509/1162142a-26d9-4913-9f15-cb5be6f1c230)

![2023-11-25_pdev_noisetest](https://github.com/aewallin/allantools/assets/909509/73410f98-1b28-4b25-b989-d949d6a2261c)

I tested against SigmaTheta (now at https://gitlab.com/fm-ltfb/SigmaTheta) and get relative errors of 1e-12 or lower, which seems ok. The remaining difference between the two programs is maybe due to hardware representation of floats in C vs. Python/numpy?

Any suggestions for what test datasets are interesting to try with PDEV?

**Comment by fmeynadier** [comment]
Interesting, thanks for the update !
I have a question concerning the plots above : for both datasets, the Taus of the PDEVs do not match the Taus of the Allantools-SigmaTheta difference, I would have expected them to be the same : am I missing something obvious ?

**Reply by aewallin** [reply]
the different taus was just a bug in my plotting script, here is an updated figure. For a length N=1000 point dataset SigmaTheta computes PDEV at an averaging factor of m=512, but AllanTools limits the maximum to m=256 here.

A practical issue is that PDEV is quite slow to compute, at least via Vernotte2020 Eqn (9) - due to the nested sums/loops. It might be possible to make the algorithm faster if the loops can be unrolled?

Using omega-counters and PDEV requires some thought and planning, for example just decimating omega-counter results from tau0 to 2*tau0 is not as straightforward as with Pi (or even Lambda-counters) - see for example Danielson2016.

As this algorithm uses phase-time input 'x' - my thinking is that it should be used with phase-measurements, or with a Pi-counter producing input data. The PDEV is then computed, and this estimates the uncertainty of an Omega-average (decimated from Omega-counter readings with the procedure in Danielson2016). Quite a bit more complex than just using and averaging with a Pi-counter.

We are looking at some applications with fiber-links (intra-lab fiber stabilization, fiber-lengths just 10-20m) and/or frequency comb locks of f0 (carrier-envelope-offset) and frep (repetition rate) where the signals might be dominated by white phase noise and Omega-counting should give a small advantage over Lambda counting (with MDEV estimates of uncetainty).

Benkler2015 is a good overall reference for counter types and the expected uncertainty for the mean measured with different types of frequency counters.

![2023-11-25_pdev_nbs14](https://github.com/aewallin/allantools/assets/909509/2e1de486-38ad-4a49-a5da-098f15cb02c9)
